File: pocketr-java.api/src/main/java/com/decrux/pocketr/api/services/currency/CurrencySeeder.java
package com.decrux.pocketr.api.services.currency;

import com.decrux.pocketr.api.entities.db.ledger.Currency;
import com.decrux.pocketr.api.repositories.CurrencyRepository;
import java.util.List;
import org.springframework.boot.ApplicationArguments;
import org.springframework.boot.ApplicationRunner;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

@Component
public class CurrencySeeder implements ApplicationRunner {

    private final CurrencyRepository currencyRepository;

    public CurrencySeeder(CurrencyRepository currencyRepository) {
        this.currencyRepository = currencyRepository;
    }

    @Override
    @Transactional
    public void run(ApplicationArguments args) {
        if (currencyRepository.count() > 0) {
            return;
        }

        List<Currency> currencies = List.of(
            currency("EUR", (short) 2, "Euro"),
            currency("USD", (short) 2, "US Dollar"),
            currency("GBP", (short) 2, "British Pound"),
            currency("JPY", (short) 0, "Japanese Yen"),
            currency("CHF", (short) 2, "Swiss Franc"),
            currency("BHD", (short) 3, "Bahraini Dinar"),
            currency("CAD", (short) 2, "Canadian Dollar"),
            currency("AUD", (short) 2, "Australian Dollar"),
            currency("SEK", (short) 2, "Swedish Krona"),
            currency("NOK", (short) 2, "Norwegian Krone"),
            currency("DKK", (short) 2, "Danish Krone"),
            currency("PLN", (short) 2, "Polish Zloty"),
            currency("CZK", (short) 2, "Czech Koruna"),
            currency("HUF", (short) 2, "Hungarian Forint"),
            currency("RON", (short) 2, "Romanian Leu")
        );

        currencyRepository.saveAll(currencies);
    }

    private static Currency currency(String code, short minorUnit, String name) {
        Currency currency = new Currency();
        currency.setCode(code);
        currency.setMinorUnit(minorUnit);
        currency.setName(name);
        return currency;
    }
}

File: pocketr-api/src/main/kotlin/com/decrux/pocketr/api/services/currency/CurrencySeeder.kt
package com.decrux.pocketr.api.services.currency

import com.decrux.pocketr.api.entities.db.ledger.Currency
import com.decrux.pocketr.api.repositories.CurrencyRepository
import org.springframework.boot.ApplicationArguments
import org.springframework.boot.ApplicationRunner
import org.springframework.core.annotation.Order
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional

@Component
@Order(0)
class CurrencySeeder(
    private val currencyRepository: CurrencyRepository,
) : ApplicationRunner {
    @Transactional
    override fun run(args: ApplicationArguments) {
        if (currencyRepository.count() > 0) return

        val currencies =
            listOf(
                Currency(code = "EUR", minorUnit = 2, name = "Euro", symbol = "EUR"),
                Currency(code = "USD", minorUnit = 2, name = "US Dollar", symbol = "$"),
                Currency(code = "GBP", minorUnit = 2, name = "British Pound", symbol = "GBP"),
                Currency(code = "JPY", minorUnit = 0, name = "Japanese Yen", symbol = "JPY"),
                Currency(code = "CHF", minorUnit = 2, name = "Swiss Franc", symbol = "CHF"),
                Currency(code = "BHD", minorUnit = 3, name = "Bahraini Dinar", symbol = "BHD"),
                Currency(code = "CAD", minorUnit = 2, name = "Canadian Dollar", symbol = "CA$"),
                Currency(code = "AUD", minorUnit = 2, name = "Australian Dollar", symbol = "A$"),
                Currency(code = "SEK", minorUnit = 2, name = "Swedish Krona", symbol = "SEK"),
                Currency(code = "NOK", minorUnit = 2, name = "Norwegian Krone", symbol = "NOK"),
                Currency(code = "DKK", minorUnit = 2, name = "Danish Krone", symbol = "DKK"),
                Currency(code = "PLN", minorUnit = 2, name = "Polish Zloty", symbol = "PLN"),
                Currency(code = "CZK", minorUnit = 2, name = "Czech Koruna", symbol = "CZK"),
                Currency(code = "HUF", minorUnit = 2, name = "Hungarian Forint", symbol = "HUF"),
                Currency(code = "RON", minorUnit = 2, name = "Romanian Leu", symbol = "RON"),
            )

        currencyRepository.saveAll(currencies)
    }
}
